R/sample_grid.R

```diff
@@ -0,0 +1,109 @@
+##'   Default = 10.
+##'   
+##' @param n_particles Number of particles. Positive Integer. Default = 100
+##'   
+##' @return List. First element is an array of trajectories, in which
+##'  first dimension is time, second is parameters chosen when sampling from 
+##'  the grid and the third dimension is alist of inputs.
+##' 
+##' @export
+##' @import furrr
+##' @importFrom utils tail
+sample_grid_scan <- function(scan_results, 
+                             n_sample_pairs = 10, 
+                             n_particles = 100) {
+  
+  # checks on args
+  assert_is(scan_results, "sircovid_scan")
+  assert_pos_int(n_sample_pairs)
+  assert_pos_int(n_particles)
+  
+  # grab the pobability matrix
+  prob <- scan_results$renorm_mat_LL
+  nr <- nrow(prob)
+  nc <- ncol(prob)
+  
+  # construct what the x and y dimensions look like
+  x_grid <- matrix(scan_results$x, nrow = nr, ncol = nc)
+  y_grid <- matrix(as.character(scan_results$y), nrow = nr, ncol = nc, byrow = TRUE)
+  
+  # draw which grid pairs are chosen
+  pairs <- sample(x =  length(prob), size = n_sample_pairs, 
+                  replace = TRUE, prob = prob)
+  
+  # what are related beta and dates
+  beta <- x_grid[pairs]
+  dates <- y_grid[pairs]
+  
+  # recreate parameters for re running
+  param_grid <- data.frame("beta" = beta, "start_date" = dates)
+  model_params <- scan_results$inputs$model_params
+  pars_obs <- scan_results$inputs$pars_obs
+  data <- scan_results$inputs$data
+  
+  
+  # Multi-core futures with furrr (parallel purrr)
+  
+  ## Particle filter outputs
+  pf_run_outputs <- furrr::future_pmap(
+    param_grid,
+    function(beta, start_date) {
+      # Edit beta in parameters
+      model_params$beta_y <- beta
+      model_params$beta_t <- 0
+      
+      # Objects used by particle filter run
+      pf_data <- particle_filter_data(data, start_date, steps_per_day= 1/model_params$dt)
+      model_run <- sircovid(params = model_params)
+      compare_func <- compare_icu(model_run, pars_obs, pf_data)
+      
+      X <- particle_filter(pf_data, model_run, compare_func, 
+                           n_particles = n_particles, save_particles = TRUE)
+    }
+  )
+  
+  # sample 1 particle from each run
+  traces <- lapply(pf_run_outputs, function(x) {
+    x$states[,,sample(n_particles, 1)]
+  })
+  
+  
+  # collapse into an array of trajectories
+  # the trajectories are different lengths in terms of dates
+  # so we will fill the arrays with NAs where needed
+  num_rows <- unlist(lapply(traces, nrow))
+  max_rows <- max(num_rows)
+  seq_max <- seq_len(max_rows)
+  trajectories <- array(NA, 
+                        dim = c(max_rows, ncol(traces[[1]]), length(traces)),
+                        dimnames = list(rownames(traces[[which.max(num_rows)]]), NULL, NULL))
+  
+  # fill the tail of the array slice
+  for (i in seq_len(length(traces))){
+    trajectories[tail(seq_max, nrow(traces[[i]])), , i] <- traces[[i]]
```

Comment:
This is so that the end of the trajectories array is populated, and the start is padded with `NA` if it's shorter than the max. If that's correct, could you clarify slightly as a comment (e.g. note the output array size, and whether the NA padding is at the start or the end)

Reply:
added comment